hubblestack/extmods/hubble_mods/bexpr.py

```diff
@@ -0,0 +1,317 @@
+          comparator:
+            type: "dict"
+            match:
+              gid: 0
+              group: "shadow"
+              uid: 0
+              user: "root"
+              mode:
+                type: "file_permission"
+                match:
+                  required_value:  "644"
+                  allow_more_strict: true
+
+check_unique_id:
+  description: 'bexpr check'
+  tag: 'ADOBE-03'
+  implementations:
+    - filter:
+        grains: 'G@osfinger:CentOS*Linux-7'
+      hubble_version: '>3 AND <7 AND <8'
+      module: bexpr
+      items:
+        - args:
+            expr: check_unique_id_1 AND check_unique_id_2
+          comparator:
+            type: boolean
+            match: True
+
+FDG Example:
+------------
+main:
+  description: 'bexpr check'
+  module: bexpr
+  args:
+    expr: check_unique_id_1 AND check_unique_id_2
+
+Mandatory parameters:
+    expr - A boolean expression referring other checks in a profile
+Multiple expressions can be provided in a single implementation under attribute: "items"
+
+Comparator compatible with this module - boolean
+
+Note: A boolean expression is evaluated in the end after all other checks have been evaluated.
+While evaluating boolean expression the result of checks referred is taken into account and no check is actually run.
+An expression is a string having following keywords:
+    AND, OR and NOT along with parenthesis '(', ')' to group the results
+
+Output of boolean expressions is same as other checks and it is classified in following categories:
+1. True - The logical expression of referenced checks evaluated to True
+2. False - The logical expression of referenced checks evaluated to False
+
+Sample Output:
+'True/False'
+"""
+import logging
+
+import hubblestack.extmods.module_runner.runner_utils as runner_utils
+from hubblestack.utils.hubble_error import HubbleCheckValidationError
+from pyparsing import infixNotation, opAssoc, Keyword, Word, alphas, ParserElement
+
+log = logging.getLogger(__name__)
+
+
+def validate_params(block_id, block_dict, chain_args=None):
+    """
+    Validate all mandatory params required for this module
+
+    :param block_id:
+        id of the block
+    :param block_dict:
+        parameter for this module
+    :param chain_args:
+        Chained argument dictionary, (If any)
+        Example: {'result': "True", 'status': True}
+
+    Raises:
+        HubbleCheckValidationError: For any validation error
+    """
+    log.debug('Module: bexpr Start validating params for check-id: {0}'.format(block_id))
+
+    error = {}
+
+    # fetch required param
+    expr = runner_utils.get_param_for_module(block_id, block_dict, 'expr')
+    if not expr:
+        error['expr'] = 'Mandatory parameter: expr not found for id: %s' % block_id
+
+    if error:
+        raise HubbleCheckValidationError(error)
+
+    log.debug('Validation success for check-id: {0}'.format(block_id))
+
+
+def execute(block_id, block_dict, result_list, chain_args=None):
```

Comment:
In the current design, there are only 3 parameters allowed. It looks like this module makes no sense in chaining. This should only be used in Audit.